In [15]:
import pandas as pd
import numpy as np
import os 

In [16]:
path = r'/Users/satoruteshima/Documents/CareerFoundry/Data Immersion 4/08-2023 Instacart Basket Analysis'

In [5]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', '01 Original Data', 'orders.csv'), index_col = False)

In [4]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', '01 Original Data', 'products.csv'), index_col = False)

## Dropping Columns

.drop (columns = [col_name] ) is going to drop the specified column

In [7]:
# Dropping 'eval_set' column from orders.csv
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

The above data is only showing the result of your funciton and NOT saving it. In order to do so, you have to use '=' to overwrite your variable! 

Be careful tho! It can’t be undone, even in the event of a mistake. 

In order to avoid re-doing the whole thing, alternatively you can assign a new vairble (df_ords_2 for example). 


#updating the variable df_ords
df_ords = df_ords.drop(columns = ['eval_set'])

OR

df_ords2 = df_ords.drop(columns = ['eval_set'])


In [8]:
df_ords = df_ords.drop(columns = ['eval_set'])

df.value_counts()  counts the number of values.
dropna = False ensures you are INCLUDING the null (Nan) value. 


In [14]:
df_ords['days_since_prior_order'].value_counts(dropna = False)


30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

.rename() is going to change the name of the column.

df.rename(columns = {'old_name' : 'new_name'}, inplace = True)

inplace = true means that you want to overwrite the original column name with this new column name rather than creating a new copy.

## Renaming Columns

In [15]:
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [17]:
df_ords.head ()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

When executing the describe() function, you can exclude these columns by changing their data type to something describe() will ignore, such as a string

## Changing Data Types

In [18]:
df_ords['order_id'] = df_ords['order_id'].astype('str')


The function above changes the data type of the "order_id" column from a number to a string

In [22]:
df_ords['order_id'].dtype

dtype('O')

O stands for string. Pandas version of 'string'

## Transposing Data

Transposing refers to turning your dataframe’s rows into columns, and vice versa.

In [4]:
#importing data from departments.csv
df_dep = pd.read_csv(os.path.join(path, '02 Data', '01 Original Data', 'departments.csv'), index_col = False)

In [27]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

Let's use python's function  T to transpose the above table!

In [28]:
#transpoting departments.csv
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [30]:
#create a new variable name for the transposing result 
df_dep_t = df_dep.T

Obviously, “0” should not be your header.

Correct this by adding something to your dataframe: an index

You can add an index to a dataframe that doesn’t have one with the following function:

In [31]:
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

Now that you have an index column, let’s create a new header for your dataframe. 

This will be done in three steps: 

1) Create a new header; 
2) Remove the first row in the dataframe;
3) Add the new header. 

In [32]:
#The iloc[] function is used for selecting data in pandas and takes the index number of a row or column (or multiple rows and columns). 
new_header = df_dep_t.iloc[0]

In [33]:
new_header

0    department
Name: department_id, dtype: object

Now we have the copy of the first row (0 row), let's delete it. 

df_dep_t_new = df_dep_t[1:]

This is telling pythong to COPY EVERYTHING FROM THE FIRST ROW ONWARD. 

In [34]:
df_dep_t_new = df_dep_t[1:]

In [36]:
df_dep_t_new.head()

0
1   frozen
2    other
3   bakery
4  produce
5  alcohol

.columns function creates the header. 

In [37]:
df_dep_t_new.columns = new_header

In [38]:
df_dep_t_new.head()

department_id department
1                 frozen
2                  other
3                 bakery
4                produce
5                alcohol

## Data Dictionary

Let’s use pandas and create a data dictionary that contains meanings for the values in the "department_id" column.

First, use the following command, which will turn your df_dep_t_new dataframe into a dictionary:



In [44]:
data_dict = df_dep_t_new.to_dict('index')

In [46]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

The to_dict() function you just used transformed your df_dep_t_new dataframe into dictionary format 

and saved it in a new variable, data_dict. 

The argument, index, tells Python to use the numbered rows as the key values for the entries in the dictionary.

In [49]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

Row 0 has a "department_id" of 19. You can use your data dictionary to determine what this stands for with the following command:

In [50]:
print(data_dict.get('19'))

{'department': 'snacks'}


## Subsetting

Subsetting refers to creating a smaller data set from a whole data set 

In [58]:
df_prods['product_name'].count()

49677

In [61]:
# create a subset for your df_prods dataframe that only contains data from the snacks department

df_snacks= df_prods[df_prods['department_id']==19]

The object['element'] syntax is called indexing. 

Indexing refers to retrieving some element from an object, be it a dataframe, list, or otherwise.

In [62]:
df_prods['department_id']==19


0         True
1        False
2        False
3        False
4        False
         ...  
49688    False
49689    False
49690    False
49691    False
49692    False
Name: department_id, Length: 49693, dtype: bool

above syntax searches all rows that has the department id of 19 and return True or False. 

In [63]:
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

An alternative way is to use loc function. While iloc functions looks for integer, loc looks for a column name. 

In [66]:
df_prods.loc[df_prods['department_id'] == 19]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

isin can include multiple values. 

In [69]:
df_prods.loc[df_prods['department_id'].isin([19, 18,17])]

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
13             14                 Fresh Scent Dishwasher Cleaner        74   
14             15                       Overnight Diapers Size 6        56   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
...           ...                                            ...       ...   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49672       49668                 Apple Cinnamon Scented Candles       101   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
13                17     6.5  
14                18    11.2  
15                19     5.2  
24                19     1.9  
...              ...     ...  
49669             19     6.9  
49670             19     5.2  
49672             17     5.6  
49675             19     3.0  
49683             19     6.0  

[10430 rows x 5 columns]

In [72]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329        1             1                   2                  8   
1  2398795        1             2                   3                  7   
2   473747        1             3                   3                 12   
3  2254736        1             4                   4                  7   
4   431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [73]:
df_ords.to_csv(os.path.join(path, '02 Data','02 Prepared Data', 'orders_wrangled.csv'))

# TASKS

### Find another identifier variable in the df_ords dataframe that doesn’t need to be included in your analysis as a numeric variable and change it to a suitable format.

In [76]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329        1             1                   2                  8   
1  2398795        1             2                   3                  7   
2   473747        1             3                   3                 12   
3  2254736        1             4                   4                  7   
4   431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [152]:
#user_id do not have to be included in my analysis as a numeric variable
df_ords['user_id'] = df_ords['user_id'].astype('str')


In [153]:
#order_number do not have to be included in my analysis as a numeric variable
df_ords['order_number'] = df_ords['order_number'].astype('str')


### Look for a variable in your df_ords dataframe with an unintuitive name and change its name without overwriting the dataframe.

In [79]:
#changing days_since_prior_order to days_since_last_order
df_ords.rename(columns = {'days_since_prior_order' : 'days_since_last_order'}, inplace = True)

In [81]:
df_ords.head(3)

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329       1             1                   2                  8   
1  2398795       1             2                   3                  7   
2   473747       1             3                   3                 12   

   days_since_last_order  
0                    NaN  
1                   15.0  
2                   21.0

### Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

In [149]:
# count the numbers of order_hour_of_day
df_ords['order_hour_of_day'].value_counts()


10    288418
11    284728
15    283639
14    283042
13    277999
12    272841
16    272553
9     257812
17    228795
18    182912
8     178201
19    140569
20    104292
7      91868
21     78109
22     61468
23     40043
6      30529
0      22758
1      12398
5       9569
2       7539
4       5527
3       5474
Name: order_hour_of_day, dtype: int64

- The businest hour is 10:00 and the least busy hour is 3:00. 

### Determine the meaning behind a value of 4 in the "department_id" column within the df_prods dataframe using a data dictionary.


In [97]:
print(data_dict.get('4'))

{'department': 'produce'}


### The sales team in your client’s organization wants to know more about breakfast item sales. Create a subset containing only the required information.


In [173]:
#locate the department_id for breakfast
df_dep_t_new.loc[df_dep_t_new['department']=='breakfast']

department_id department
14             breakfast

In [174]:
#create a subset containing breakfast products
df_breakfast = df_prods.loc[df_prods['department_id']==14]

### They’d also like to see details about products that customers might use to throw dinner parties. Your task is to find all observations from the entire dataframe that include items from the following departments: alcohol, deli, beverages, and meat/seafood. You’ll need to present this subset to your client.


In [127]:
# finding department_id for the target group
df_dep_t_new.loc[df_dep_t_new['department'].isin(['pantry', 'deli', 'beverages', 'meat seafood'])]


department_id    department
7                 beverages
12             meat seafood
13                   pantry
20                     deli

In [128]:
# assigning target_departments to a variable
target_departments = [5, 20, 7, 12]

In [175]:
dinner_parties = df_prods.loc[df_prods['department_id'].isin(target_departments)]

In [176]:
dinner_parties

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

### It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have?

- 7650 rows

### Someone from the data engineers team in Instacart thinks they’ve spotted something strange about the customer with a "user_id" of “1.” Extract all the information you can about this user.

In [155]:
user_id_1 = df_ords.loc[df_ords['user_id']=='1']

In [156]:
user_id_1

order_id user_id order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1            1                   2                  8   
1   2398795       1            2                   3                  7   
2    473747       1            3                   3                 12   
3   2254736       1            4                   4                  7   
4    431534       1            5                   4                 15   
5   3367565       1            6                   2                  7   
6    550135       1            7                   1                  9   
7   3108588       1            8                   1                 14   
8   2295261       1            9                   1                 16   
9   2550362       1           10                   4                  8   
10  1187899       1           11                   4                  8   

    days_since_last_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0  
5                    19.0  
6                    20.0  
7                    14.0  
8                     0.0  
9                    30.0  
10                   14.0

In [157]:
user_id_1.describe()

orders_day_of_week  order_hour_of_day  days_since_last_order
count           11.000000          11.000000              10.000000
mean             2.636364          10.090909              19.000000
std              1.286291           3.477198               9.030811
min              1.000000           7.000000               0.000000
25%              1.500000           7.500000              14.250000
50%              3.000000           8.000000              19.500000
75%              4.000000          13.000000              26.250000
max              4.000000          16.000000              30.000000

### You also need to provide some details about this user’s behavior. What basic stats can you provide based on the information you have?

- This user has made 11 orders in total. 
- The longest days they waited for a new oder is 29 days. 
- Order ID '2295261'  was made immediately after the last order. 
- They most frequently make an order on Thursday. 
- They most likely make an order in the morning from 8:00 to 10:00. 

### Export your df_ords dataframe as “orders_wrangled.csv” in your “Prepared Data” folder.

In [158]:
df_ords.to_csv(os.path.join(path, '02 Data','02 Prepared Data', 'orders_wrangled.csv'))

In [159]:
df_dep_t_new.to_csv(os.path.join(path, '02 Data','02 Prepared Data', 'departments_wrangled.csv'))